### PR TITLE
Inmem Storage: Observe single partition from multiple partitions.

### PR DESCRIPTION
**Pull request by dmvk**

### CODE REVIEW

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorageTest.java [hunk 1]
```diff
@@ -631,34 +631,93 @@ public boolean onError(Throwable error) {
     assertEquals(3, observeHandle.getCurrentOffsets().size());
   }
 
-  @Test(expected = UnsupportedOperationException.class)
+  @Test
+  public void testObserveSinglePartitionOutOfMultiplePartitions() throws InterruptedException {
```
Comment (je-ik) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorageTest.java:635]: This test looks it would pass even when all the data would be in single partition. Can we add a check for the keys that we process?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorageTest.java [hunk 2]
```diff
@@ -631,34 +631,93 @@ public boolean onError(Throwable error) {
...
+              public void onRepartition(OnRepartitionContext context) {
+                assertEquals(numPartitions, context.partitions().size());
+              }
+
+              @Override
+              public boolean onNext(StreamElement ingest, OnNextContext context) {
+                partitionHistogram.merge(context.getPartition(), 1L, Long::sum);
+                context.confirm();
+                elementsReceived.countDown();
+                return elementsReceived.getCount() > 0;
```
Comment (je-ik) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorageTest.java:663]: Maybe here, we can read the key and check that it satisfies some condition, like `Integer.parseInt(ingest.getKey().split("_")) % 3 == 0`

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 3]
```diff
@@ -537,25 +545,27 @@ private void handleFlushDataBaseOnPosition(
           scheduler.scheduleAtFixedRate(
               onIdle, IDLE_FLUSH_TIME, IDLE_FLUSH_TIME, TimeUnit.MILLISECONDS);
       onIdleRef.set(onIdleFuture);
-      AtomicReference<StreamElement> lastConsumed = new AtomicReference<>();
+      final ConcurrentMap<Partition, StreamElement> lastConsumedPerPartition =
```
Comment (je-ik) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:548]: This probably doesn't need ConcurrentMap, because there is lock held when accessing it.

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 4]
```diff
@@ -307,7 +308,12 @@ public ObserveHandle observePartitions(
         Position position,
         boolean stopAtCurrent,
         LogObserver observer) {
-      return observe(null, position, stopAtCurrent, observer);
+      return observe(
+          null,
+          position,
+          partitions.stream().map(ConsumedOffset::empty).collect(Collectors.toList()),
+          stopAtCurrent,
+          observer);
```
Comment (je-ik) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:316]: Why this change? It should work as before, right? The call to `observe(null, position, stopAtCurrent, observer)` is public and thus should work.

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 5]
```diff
@@ -1062,7 +1070,7 @@ public DataAccessor createAccessor(
             .map(v -> Integer.parseInt(v.toString()))
             .orElse(1);
 
-    final Partitioner partitioner = new KeyAttributePartitioner();
+    final Partitioner partitioner = new KeyPartitioner();
```
Comment (je-ik) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:1073]: Why did we change the default? The default in other commit-logs is KeyAttribute partitioner.

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 6]
```diff
@@ -467,7 +473,10 @@ public void waitUntilReady() {
                   () ->
                       handleFlushDataBaseOnPosition(
                           position,
-                          initialOffsets,
+                          initialOffsets
+                              .stream()
+                              .map(ConsumedOffset::getPartition)
+                              .collect(Collectors.toSet()),
```
Comment (je-ik) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:479]: :+1:

**Review by je-ik — CHANGES_REQUESTED** [review]
Cool, I have some minor comments, the biggest concern is about changing the default partitioner. We should adhere to the defaults used by "standard" commit-logs. If the change is motivated by the test, we should either change the test, or change the partitioner only inside the test.

**Comment by je-ik** [comment]
@dmvk will you have time to finish this?

**Comment by dmvk** [comment]
Yes, I'll be finishing Flink integration within next two weeks and this is needed for test cases to work properly.

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 5 above)
Comment (dmvk) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:1073]: Let me see if change breaks anything. There was a motivation around simpler partition for tests (where you need to know, where your element ends up, before you write it), but that's something we can work around with proper tooling.

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 3 above)
Comment (dmvk) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:548]: seems correct 👍

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorageTest.java (hunk 1 above)
Comment (dmvk) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorageTest.java:635]: fixed

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 4 above)
Comment (dmvk) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:316]: Because we need to take "partitions" parameter into account? (previous implementation contains bug and was not cover with tests)

**Comment by dmvk** [comment]
@je-ik Ready for another pass

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 4 above)
Comment (je-ik) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:361]: Ah, yep. :)

**Review by je-ik — APPROVED** [review]
LGTM! Thanks!
